File: day3_crossed_wires/python/day3_crossed_wires.py
#!/bin/python

import sys
import io
from enum import Enum

class Direction(Enum):
    up = 1
    down = 2
    left = 3
    right = 4


class Move:
    def __init__(self, direction, distance):
        self.direction = direction
        self.distance = distance

    def __repr__(self):
        return "{} {}".format(self.direction, self.distance)

    def apply(self, x0, y0):
        if self.direction == Direction.up:
            return (x0, y0 - self.distance)
        elif self.direction == Direction.down:
            return (x0, y0 + self.distance)
        elif self.direction == Direction.left:
            return (x0 - self.distance, y0)
        elif self.direction == Direction.right:
            return (x0 + self.distance, y0)

    def to_segment(self, x0, y0):
        if self.direction == Direction.up:
            return WireSegment(x0, y0 - self.distance, x0, y0)
        elif self.direction == Direction.down:
            return WireSegment(x0, y0, x0, y0 + self.distance)
        elif self.direction == Direction.left:
            return WireSegment(x0 - self.distance, y0, x0, y0)
        elif self.direction == Direction.right:
            return WireSegment(x0, y0, x0 + self.distance, y0)


class WireSegment:
    def __init__(self, x0, y0, x1, y1):
        self.x0 = x0
        self.y0 = y0
        self.x1 = x1
        self.y1 = y1

    def is_horizontal(self):
        return self.y0 == self.y1

    def is_vertical(self):
        return self.x0 == self.x1

    def crossing_point(self, other_wire_segment):
        if self.is_horizontal():
            if other_wire_segment.is_horizontal():
                return None
            elif other_wire_segment.x0 >= self.x0 and other_wire_segment.x0 <= self.x1 and self.y0 >= other_wire_segment.y0 and self.y0 <= other_wire_segment.y1:
                return (other_wire_segment.x0, self.y0)
            else:
                return None
        else: # this segment is vertical
            if other_wire_segment.is_vertical():
                return None
            elif other_wire_segment.y0 >= self.y0 and other_wire_segment.y0 <= self.y1 and self.x0 >= other_wire_segment.x0 and self.x0 <= other_wire_segment.x1:
                return (self.x0, other_wire_segment.y0)
            else:
                return None


def manhattan_distance_from_origin(x, y):
    return abs(x) + abs(y)


def parse_direction(c):
    if c == 'U':
        return Direction.up
    elif c == 'D':
        return Direction.down
    elif c == 'L':
        return Direction.left
    elif c == 'R':
        return Direction.right


def parse_move(m):
    return Move(parse_direction(m[0]), int(m[1:]))


def closest(d1, d2):
    if not d1:
        return d2
    elif not d2:
        return d1
    elif d1 < d2:
        return d1
    else:
        return d2


def find_closest_crossing(wire1_moves, wire2_moves):
    wire1_x, wire1_y = 0, 0
    best_crossing_distance = None
    for wire1_move in wire1_moves:
        wire1_segment = wire1_move.to_segment(wire1_x, wire1_y)
        wire2_x, wire2_y = 0, 0
        for wire2_move in wire2_moves:
            wire2_segment = wire2_move.to_segment(wire2_x, wire2_y)
            c = wire1_segment.crossing_point(wire2_segment)
            if c:
                (cx, cy) = c
                best_crossing_distance = closest(best_crossing_distance, manhattan_distance_from_origin(cx, cy))
            (wire2_x, wire2_y) = wire2_move.apply(wire2_x, wire2_y)
        (wire1_x, wire1_y) = wire1_move.apply(wire1_x, wire1_y)
    return best_crossing_distance


file_path = sys.argv[1]
wire1_moves = []
wire2_moves = []
with io.open(file_path, "r") as f:
    wire1_moves = [parse_move(m) for m in f.readline().split(",")]
    wire2_moves = [parse_move(m) for m in f.readline().split(",")]

closest_crossing_distance = find_closest_crossing(wire1_moves, wire2_moves)

print(closest_crossing_distance)

File: day3_crossed_wires/rust/day3_crossed_wires.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};


fn main() {
    let args: Vec<String> = env::args().collect();
    let file_path: &String = &args[1];

    let f = File::open(file_path).expect("Unable to open file");
    let f = BufReader::new(f);

    let mut fli = f.lines();
    let wire1_moves = load_moves(&fli.next().expect("No wire 1 moves").expect("No wire 1 moves"));
    let wire2_moves = load_moves(&fli.next().expect("No wire 2 moves").expect("No wire 2 moves"));

    println!("{}", find_closest_crossing(&wire1_moves, &wire2_moves));
}


fn load_moves(moves_csv: &str) -> Vec<Move> {
    let mut ret = Vec::<Move>::new();
    for s in moves_csv.split(",") {
        ret.push(parse_move(&s));
    }
    ret
}


enum Direction {
    Up,
    Down,
    Left,
    Right
}


struct Move {
    direction: Direction,
    distance: i32
}


impl Move {

    fn apply(&self, pos: &WirePos) -> WirePos {
        match self.direction {
            Direction::Up =>    WirePos{x: pos.x, y: pos.y - self.distance},
            Direction::Down =>  WirePos{x: pos.x, y: pos.y + self.distance},
            Direction::Left =>  WirePos{x: pos.x - self.distance, y: pos.y},
            Direction::Right => WirePos{x: pos.x + self.distance, y: pos.y},
        }
    }

    fn to_segment(&self, pos: &WirePos) -> WireSegment {
        match self.direction {
            Direction::Up =>    WireSegment{x0: pos.x, y0: pos.y - self.distance, x1: pos.x, y1: pos.y},
            Direction::Down =>  WireSegment{x0: pos.x, y0: pos.y, x1: pos.x, y1: pos.y + self.distance},
            Direction::Left =>  WireSegment{x0: pos.x - self.distance, y0: pos.y, x1: pos.x, y1: pos.y},
            Direction::Right => WireSegment{x0: pos.x, y0: pos.y, x1: pos.x + self.distance, y1: pos.y},
        }
    }
}


struct WireSegment {
    x0: i32,
    y0: i32,
    x1: i32,
    y1: i32,
}

impl WireSegment {
    fn is_horizontal(&self) -> bool {
        self.y0 == self.y1
    }

    fn crossing_point(&self, other_wire_segment: &WireSegment) -> Option::<(i32, i32)> {
        match (self.is_horizontal(), other_wire_segment.is_horizontal()) {
            (true, true) => None,
            (false, false) => None,
            (true, false) =>
                if other_wire_segment.x0 >= self.x0 && other_wire_segment.x0 <= self.x1 && self.y0 >= other_wire_segment.y0 && self.y0 <= other_wire_segment.y1 {
                    Some((other_wire_segment.x0, self.y0))
                } else {
                    None
                },
            (false, true) =>
                if other_wire_segment.y0 >= self.y0 && other_wire_segment.y0 <= self.y1 && self.x0 >= other_wire_segment.x0 && self.x0 <= other_wire_segment.x1 {
                    Some((self.x0, other_wire_segment.y0))
                } else {
                    None
                },
        }
    }
}


struct WirePos {
    x: i32,
    y: i32,
}


fn parse_direction(c: char) -> Direction {
    match c {
        'U' => Direction::Up,
        'D' => Direction::Down,
        'L' => Direction::Left,
        'R' => Direction::Right,
        _ => panic!("Bad direction"),
    }
}


fn parse_move(m : &str) -> Move {
    Move{
        direction: parse_direction(m.chars().next().expect("Empty move")),
        distance: str::parse::<i32>(&m[1..]).expect("Bad step count")
    }
}


fn minimal(d1: Option<i32>, d2: Option<i32>) -> Option<i32> {
    match (d1, d2) {
        (None, None) => None,
        (None, Some(y)) => Some(y),
        (Some(x), None) => Some(x),
        (Some(0), Some(y)) => Some(y),
        (Some(x), Some(0)) => Some(x),
        (Some(x), Some(y)) => if x < y {Some(x)} else {Some(y)},
    }
}


fn manhattan_distance_from_origin(x: i32, y: i32) -> i32 {
    x.abs() + y.abs()
}


fn find_closest_crossing(wire1_moves: &Vec::<Move>, wire2_moves: &Vec::<Move>) -> i32 {
    let mut wire1_pos = WirePos{x: 0, y: 0};
    let mut best_crossing_distance: Option::<i32> = None;
    for wire1_move in wire1_moves {
        let wire1_segment = wire1_move.to_segment(&wire1_pos);
        let mut wire2_pos = WirePos{x: 0, y: 0};
        for wire2_move in wire2_moves {
            let wire2_segment = wire2_move.to_segment(&wire2_pos);
            match wire1_segment.crossing_point(&wire2_segment) {
                Some((cx, cy)) => {
                    best_crossing_distance = minimal(best_crossing_distance, Some(manhattan_distance_from_origin(cx, cy)));
                },
                None => (),
            }
            wire2_pos = wire2_move.apply(&wire2_pos);
        }
        wire1_pos = wire1_move.apply(&wire1_pos);
    }
    best_crossing_distance.expect("No crossing found!")
}
